# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [3]:
def najdi_prvocisla(pocet, prvocisla = []):
    if len(prvocisla)>=pocet:
        return prvocisla[:pocet]
    n = 2
    while len(prvocisla) < pocet:
        check = True
        for i in range(2, n):
            if(n%i == 0):
                check = False
                break
        if check:
            prvocisla.append(n)
        n+=1
    return prvocisla[:pocet]

print(najdi_prvocisla(10))
print(najdi_prvocisla(20))
print(najdi_prvocisla(5))

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
[2, 3, 5, 7, 11]



# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [7]:
%%writefile ahoj.py

def odmocnina(a, n):
    x=a
    for i in range(n):
        x=((a/x)+x)/2
    return x

def pi_v1(n):
    v_n = odmocnina(1-(0.5)**2,100)
    b_n = 1
    a = 1
    for i in range(n):
        b_2n = odmocnina(((b_n/2)**2)+((1-v_n)**2),100)
        v_2n = odmocnina(1-((b_2n/2)**2),100)
        b_n = b_2n
        v_n = v_2n
    return 6*(2**n)*b_2n*v_2n/2

def pi_v2(n):
    suma = 0
    a = 1/(2**4)
    for i in range(1, n+1):
        suma+=a/((2*i)+1)
        a = a * ((2*(i+1)-3)/(2*(i+1)))*(1/(2**2))
    return 12*((-odmocnina(3,100000)/8)+0.5-suma)


Overwriting ahoj.py


In [9]:
def odmocnina(a, n):
    x=a
    for i in range(n):
        x=((a/x)+x)/2
    return x

def pi_v1(n):
    v_n = odmocnina(1-(0.5)**2,100)
    b_n = 1
    a = 1
    for i in range(n):
        b_2n = odmocnina(((b_n/2)**2)+((1-v_n)**2),100)
        v_2n = odmocnina(1-((b_2n/2)**2),100)
        b_n = b_2n
        v_n = v_2n
    return 6*(2**n)*b_2n*v_2n/2

def pi_v2(n):
    suma = 0
    a = 1/(2**4)
    for i in range(1, n+1):
        suma+=a/((2*i)+1)
        a = a * ((2*(i+1)-3)/(2*(i+1)))*(1/(2**2))
    return 12*((-odmocnina(3,100000)/8)+0.5-suma)

In [13]:
%time pi_v1(20)
print("------------------------")
%timeit pi_v2(20)


3.14159265358927
------------------------
3.1415926535897936


In [14]:
%%time
odmocnina(1000,10000)
pi_v1(50)
pi_v2(50)

CPU times: user 52.5 ms, sys: 3.85 ms, total: 56.3 ms
Wall time: 145 ms


3.1415926535897936

In [15]:
%%timeit
odmocnina(1000,100)
pi_v1(10)
pi_v2(10)

55.5 ms ± 7.43 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)



# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [16]:
import math
n = 10
piv1 = pi_v1(n)
piv2 = pi_v2(n)
psc1 = math.floor(- math.log10(abs(math.pi - piv1)))
psc2 = math.floor(- math.log10(abs(math.pi - piv2)))
print(f"n = {n}, pi1 = {piv1}, pocet spravnych cifer = {psc1}")
print(f"n = {n}, pi2 = {piv2}, pocet spravnych cifer = {psc2}")

n = 10, pi1 = 3.141592105999271, pocet spravnych cifer = 6
n = 10, pi2 = 3.1415926542141035, pocet spravnych cifer = 9



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.


In [29]:
def genfib(n):
    a1 = 0
    a2 = 1
    for i in range(n):
        yield a1
        a3 = a1 + a2
        a1 = a2
        a2 = a3

gen = genfib(10)
for i in range(10):
    print(next(gen))



0
1
1
2
3
5
8
13
21
34



# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).


In [57]:
print([i for i in range(1, 11)])
print([i**2 for i in range(1, 11)])
print([i for i in range(2, 101) if all(i % j != 0 for j in range(2, int(i**0.5)+1))])


[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]



# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

In [30]:
%%writefile ahojda.txt
jablko 2
hruska 2
avokado 4
merunka 6
ananas 1
jahoda 2
バナナ 3

Writing ahojda.txt


In [38]:
from math import sqrt
with open("ahojda.txt", "r") as soubor:
    for i, radek in enumerate(soubor):
        print(f"{radek.rstrip()}")

with open("druhy-soubor.txt", "a") as soubor:
    soubor.write(f"Mam rad f stringy!\n")
    for i in range(1,11):
        soubor.write(f"Cislo = {i}, druha mocnina = {i**2}, druha odmocnina = {sqrt(i)}\n")

jablko 2
hruska 2
avokado 4
merunka 6
ananas 1
jahoda 2
バナナ 3


# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [9]:
# kostra k doplnění, místa označená TODO doplňte

class Operator:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass

    def simplify(self):
        pass
    
    def __add__(self, other):
        return Addition(self, other)
    
    def __sub__(self, other):
        return Addition(self, Multiplication(other, cst(-1)))
    
    def __mul__(self, other):
        return Multiplication(self, other)
    
class BinaryOperator(Operator):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Operator):
    def __init__(self, argument):
        self.argument = argument
    def __str__(self):
        return type(self).__name__ + "(" + str(self.argument) + ")"

class Addition(BinaryOperator):
    def __str__(self):
        return f"({self.levy_argument}+{self.pravy_argument})"
    def diff(self):
        return Addition(self.levy_argument.diff(), self.pravy_argument.diff())
    def simplify(self):
        if isinstance(self.levy_argument, cst) and self.levy_argument.value == 0:
            return self.pravy_argument
        if isinstance(self.pravy_argument, cst) and self.pravy_argument.value == 0:
            return self.levy_argument
    
class Multiplication(BinaryOperator):
    def __str__(self):
        return f"({self.levy_argument}*{self.pravy_argument})"
    def diff(self):
        return Addition(Multiplication(self.levy_argument.diff(), self.pravy_argument), Multiplication(self.levy_argument, self.pravy_argument.diff()))
    def simplify(self):
        if isinstance(self.levy_argument, cst) and self.levy_argument.value == 0:
            return self.levy_argument
        if isinstance(self.pravy_argument, cst) and self.pravy_argument.value == 0:
            return self.pravy_argument
        if isinstance(self.levy_argument, cst) and self.levy_argument.value == 1:
            return self.pravy_argument
        if isinstance(self.pravy_argument, cst) and self.pravy_argument.value == 1:
            return self.levy_argument
        if isinstance(self.levy_argument, cst) and isinstance(self.pravy_argument, cst):
            return cst(self.levy_argument.value * self.pravy_argument.value)
        return self
    
class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return str(self.value)
    def diff(self):
        return cst(0)

class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        return 'x'
    def diff(self):
        return cst(1)
    
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        return f"{self.argument}^{self.exponent}"
    def diff(self):
        return Multiplication(self.exponent, mocnina(self.argument,self.exponent-1))
    
class sin(UnaryOperator):
    def diff(self):
        return f"(cos({self.argument})*({self.argument.diff()})"

class cos(UnaryOperator):
    def diff(self):
        return f"(-sin({self.argument})*({self.argument.diff()})"

class exp(UnaryOperator):
    def diff(self):
        return f"(exp({self.argument}))*({self.argument.diff()})"
    
class ln(UnaryOperator):
    def diff(self):
        return f"(1/x)*({self.argument.diff()})"
    
class frc(UnaryOperator):
    def __str__(self):
        return f"(1/{self.argument})"
    def diff(self):
        return f"(1/(({self.argument})^2))*(-1)*({self.argument.diff()})"

In [10]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())

print((cst(2)*x+cst(1)))
print((cst(2)*x+cst(1)).diff())
f4 = (cst(2)*x+cst(1)).diff()
print(f4.simplify())

#simplify se mi nepodarilo zprovoznit

(sin((x+cos((x*x))))+1)
((cos((x+cos((x*x))))*((1+(-sin((x*x))*(((1*x)+(x*1)))))+0)
((2*x)+1)
(((0*x)+(2*1))+0)
((0*x)+(2*1))


můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [63]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


(((x^2+(2)*(x))+1)+(2)*(x^2))
((((2)*(x^1)+((0)*(x)+(2)*(1)))+0)+((0)*(x^2)+(2)*((2)*(x^1))))


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [71]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())


((x^2+(2)*(x))+exp(x))
(((2)*(x^1)+((0)*(x)+(2)*(1)))+(exp(x))*(1))


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...